src/garage/torch/policies/context_conditioned_policy.py

```diff
@@ -0,0 +1,271 @@
+        task_z = [z.repeat(b, 1) for z in task_z]
+        task_z = torch.cat(task_z, dim=0)
+
+        # run policy, get log probs and new actions
+        obs_z = torch.cat([obs, task_z.detach()], dim=1)
+        dist = self._policy(obs_z)
+        pre_tanh, actions = dist.base_dist.rsample_with_pre_tanh_value()
+        log_pi = dist.base_dist.log_prob(value=actions,
+                                         pre_tanh_value=pre_tanh)
+        log_pi = log_pi.unsqueeze(1)
+        mean = dist.mean.to('cpu').detach().numpy()
+        log_std = (dist.variance**.5).log().to('cpu').detach().numpy()
+
+        return (actions, mean, log_std, log_pi, pre_tanh), task_z
+
+    def get_action(self, obs):
+        """Sample action from the policy, conditioned on the task embedding.
+
+        Args:
+            obs (torch.Tensor): Observation values.
+
+        Returns:
+            torch.Tensor: Output action values.
+
+        """
+        z = self.z
+        obs = tu.from_numpy(obs[None])
+        obs_in = torch.cat([obs, z], dim=1)
+        action, info = self._policy.get_action(obs_in)
+        action = np.squeeze(action, axis=0)
+        info['mean'] = np.squeeze(info['mean'], axis=0)
+        return action, info
+
+    def compute_kl_div(self):
+        """Compute KL( q(z|c) || p(z) ).
+
+        Returns:
+            float: KL( q(z|c) || p(z) ).
+
+        """
+        prior = torch.distributions.Normal(tu.zeros(self._latent_dim),
+                                           tu.ones(self._latent_dim))
+        posteriors = [
+            torch.distributions.Normal(mu, torch.sqrt(var)) for mu, var in zip(
+                torch.unbind(self.z_means), torch.unbind(self.z_vars))
+        ]
+        kl_divs = [
+            torch.distributions.kl.kl_divergence(post, prior)
+            for post in posteriors
+        ]
+        kl_div_sum = torch.sum(torch.stack(kl_divs))
+        return kl_div_sum
+
+    def reset(self, dones=None):
+        """Reset the environment.
+
+        Args:
+            dones (numpy.ndarray): Reset values.
+
+        """
+
+    def log_diagnostics(self, eval_statistics):
+        """Log data about encodings to eval_statistics dictionary.
+
+        Args:
+            eval_statistics (dict): Dictionary for logging info.
+
+        """
+        z_mean = np.mean(np.abs(tu.to_numpy(self.z_means[0])))
+        z_sig = np.mean(tu.to_numpy(self.z_vars[0]))
+        eval_statistics['ZMeanEval'] = z_mean
+        eval_statistics['ZVarianceEval'] = z_sig
+
+    @property
+    def networks(self):
+        """Return context_encoder and policy.
+
+        Returns:
+            list: Encoder and policy networks.
+
+        """
+        return [self._context_encoder, self._policy]
+
+    @property
+    def context(self):
+        """Return context.
+
+        Returns:
+            torch.Tensor: Context values.
+
+        """
+        return self._context
+
+
+def _product_of_gaussians(mus, sigmas_squared):
```

Comment:
hmm perhaps this can go into garage.torch.utils?

Reply:
Done.